Made ElementDeletionGeneratorBase delete elems that have already deleted elems as an interior parent

socratesgorilla opened the pull request:
Closes #21983

Commit:
Made ElementDeletionGeneratorBase delete elems that had already deleted elems as an interior parent #21983

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1187220/) on 100af79 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22241/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22241/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ae03f3895eb5ff98b73b5cf79b5d20cb6ecf6c02`

Commit:
Updated tests and applied code review suggestions

Modified tests, reverted conditional check

Update test/tests/meshgenerators/block_deletion_generator/delete_interior_parents.i

Changed test to exodiff, added check for lowerd elems

Changes from review suggestions

Added alert in documentation

Apply suggestions from code review

Co-Authored-By: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1187278/) on 3add8fd wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22241/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1187530/) on 2e1e73a wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>ae03f3</th>
      <th colspan=4>#22241 2e1e73</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.28%</td>
      <td>84.28%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>79971</td>
      <td>79975</td>
      <td>+4</td>
      <td>5</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14917</td>
      <td>14918</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22241/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22241/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/richards)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22241/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on framework/src/meshgenerators/ElementDeletionGeneratorBase.C:
```diff
@@ -226,6 +235,7 @@ ElementDeletionGeneratorBase::generate()
    * For now, we'll call contract and notify the SetupMeshComplete
    * Action that we need to re-prepare the mesh.
    */
+
```
```suggestion
```

GiudGiud commented on framework/src/meshgenerators/ElementDeletionGeneratorBase.C:
```diff
@@ -52,6 +52,15 @@ ElementDeletionGeneratorBase::generate()
+  // Set to be merged into deleteable elems, used to avoid derefencing a dangling pointer
+  std::set<Elem *> deleteable_parents;
+
+  // check for dangling interior parents
+  for (auto & elem : mesh->element_ptr_range())
```
this is costing you another loop on all the mesh

GiudGiud replied:
ok so check the element dimension before doing a search. this will reduce the complexity from:
N_mesh * N_elem_deleted to
N_lower-d-mesh * N_elem_deleted

and that should be enough

GiudGiud commented on framework/src/meshgenerators/ElementDeletionGeneratorBase.C:
```diff
@@ -52,6 +52,15 @@ ElementDeletionGeneratorBase::generate()
       deleteable_elems.insert(elem);
```
when you could just loop over the child of the element here

GiudGiud replied:
well it s not that simple

GiudGiud commented on test/tests/meshgenerators/block_deletion_generator/delete_interior_parents.i:
```diff
@@ -0,0 +1,72 @@
+[Mesh]
+  [twoquad]
+    type = GeneratedMeshGenerator
+    dim = 2
+    nx = 2
+    xmax = 2
```
```suggestion
    dim = 2
    nx = 3
    ny = 4
    xmax = 2
```
to make sure we hit more cases

GiudGiud replied:
a 2x2 boundary does not have a fully interior element

GiudGiud commented on test/tests/meshgenerators/block_deletion_generator/tests:
```diff
@@ -33,6 +33,13 @@
       detail = 'blocks and assigning the new boundary even if the input mesh is not prepared.'
     []
+    [delete_interior_parents]
+      type = 'RunApp'
```
this should be an exodiff.
We need to make sure we get the right result out of this

GiudGiud commented on framework/src/meshgenerators/ElementDeletionGeneratorBase.C:
```diff
@@ -52,6 +52,16 @@ ElementDeletionGeneratorBase::generate()
       deleteable_elems.insert(elem);
   }
 
+  // Set to be merged into deleteable elems, used to avoid derefencing a dangling pointer
+  std::set<Elem *> deleteable_parents;
+
+  // check for dangling interior parents
+  for (auto & elem : mesh->element_ptr_range())
+    if (elem->dim() == mesh->mesh_dimension() - 1 &&
+        deleteable_elems.count(elem->interior_parent()) > 0)
+      deleteable_parents.insert(elem);
+
+  deleteable_elems.insert(deleteable_parents.begin(), deleteable_parents.end());
```
why cant you insert in deleteable elements?

also isnt it deleteable_child rather than parents here?

GiudGiud commented on test/tests/meshgenerators/block_deletion_generator/tests:
```diff
@@ -33,6 +33,13 @@
+      input = 'delete_interior_parents.i'
+      exodiff = 'delete_interior_parents_out.e'
+      recover = false
+      detail = 'blocks and properly removing elements with a deleted element as their interior parent.'
```
```suggestion
      detail = 'blocks and properly removing lower dimensional elements with a deleted element as their interior parent.'
```

GiudGiud commented on test/tests/meshgenerators/block_deletion_generator/delete_interior_parents.i:
```diff
@@ -0,0 +1,76 @@
+    bottom_left = '1 0 0'
+    top_right = '2 1 0'
+  []
+  [sideset]
+    type = SideSetsBetweenSubdomainsGenerator
+    input = subdomain
+    primary_block = 0
+    paired_block = 1
+    new_boundary = bar
+  []
+  [edge_elem]
+    type = LowerDBlockFromSidesetGenerator
+    input = sideset
+    sidesets = bar
+    new_block_id = 2
+  []
```
duplicate those please with foo for the sideset from 1 to 0 ((instead of 0 to 1)

I want to see if those lowerD elements survive as expected

GiudGiud reviewed (changes requested):
Add a note in the documentation about this
For example

!alert note
Lower-dimensional elements created from the deleted elements will also be removed

socratesgorilla commented:
Alert note is in there

GiudGiud commented on framework/src/meshgenerators/ElementDeletionGeneratorBase.C:
```diff
@@ -52,10 +52,15 @@ ElementDeletionGeneratorBase::generate()
+    if (elem->dim() == mesh->mesh_dimension() - 1 &&
+        deleteable_elems.count(elem->interior_parent()) > 0)
+      deleteable_elems.insert(elem);
+      /**
+       * If we are in parallel we'd better have a consistent idea of what
+       * should be deleted.  This can't be checked cheaply.
+       */
```
```suggestion
  /**
   * If we are in parallel we'd better have a consistent idea of what
   * should be deleted.  This can't be checked cheaply.
   */
```

GiudGiud commented:
clean up the commit history

we dont want a merge commit

GiudGiud commented:
what you can do is an interactive rebase, remove the merge and squash the unimportant ones

socratesgorilla commented on test/tests/meshgenerators/block_deletion_generator/delete_interior_parents.i:
```diff
@@ -0,0 +1,95 @@
+    sidesets = bar
+    new_block_id = 2
+  []
+  [delete2]
+    type = BlockDeletionGenerator
+    input = edge_elem2
+    block = 0
+  []
```
```suggestion
```

socratesgorilla commented:
I don't even know how that merge commit got in there, it shouldn't be. I don't remember merging.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1189901/) on 3ba2b0a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22241/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22241/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 71dc08770e00d3f37c618341d5766a662819a0a9`

Commit:
Fixed conflicts with existing branch and clang-formatted

Commit:
Made ElementDeletionGeneratorBase delete elems that had already deleted elems as an interior parent #21983

socratesgorilla commented on test/tests/meshgenerators/block_deletion_generator/delete_interior_parents.i:
```diff
@@ -0,0 +1,90 @@
+    type = LowerDBlockFromSidesetGenerator
+    input = sideset2
+    sidesets = bar
+    new_block_id = 2
```
Well, the lowerD block is, to the block_deletion_generator, just another block, the fact that someone made a new block with a new block id using the LowerDBlock generator shouldn't have any bearing. It doesn't make any sense that it wouldn't be able to delete the block, the selection of block is essentially arbitrary.

As long as it properly discerns between the right elements, there shouldn't be an issue, right? As far as I know that's what is happening.

GiudGiud replied:
Let's say we have 
block 1, 2 higher D blocks
block 3 tied to elements in block 1, lower D, on sideset 1 between 1 & 2
block 4 tied to elements in block 2, lower D, on sideset 2 between 2 & 1

I d want the BlockDeletionGenerator to be able to:
- delete block 3 when deleting block 1 (this might be controversial actually)
- delete block 4 when deleting block 2 (same)
- delete block 3 or 4 only

socratesgorilla replied:
Do you want me to test for that behavior?

GiudGiud replied:
I d like the test that deletes block 1 & 3 also contain block 4 and does not delete it.

You are pretty much there, you just need to get the MGs set up properly and fix the block ids created I think

WilkAndy replied:
As far as i'm concerned, the most important thing is to properly document this (eg, with examples, similar to @GiudGiud 's above).  BUT, if i had my choice, i would not want to always delete block 3 when deleting block 1.  I don't see why you'd always want to do that, and in my situation you'd almost always want to not delete block 3.  I'd prefer to specify block = '1 3' if i wanted to delete both blocks.   But, as i said, the documentation is more important than the final functionality.

GiudGiud replied:
The problem is that:
- if you specify '1 3' you already delete both blocks currently
- if you only specify 1, it seg faults because of elements missing their interior parent.
I agree documenting the behavior is important here

socratesgorilla replied:
Newest push deletes block 1 and 3 and doesnt delete block 4.

Commit:
Modified delete_interior_parents.i

Commit:
Changed gold file

GiudGiud reviewed (approved):
This looks good.
If users want to leave the lower-D elements in the future they ll just add a boolean and some more logic to fix the interior parents pointers.

Thanks for the fix!
